File: app/src/main/java/com/digis/android/task/implementation/java/data/BaseRepository.java
package com.digis.android.task.implementation.java.data;

import android.app.Application;

import com.digis.android.task.implementation.java.data.source.remote.RemoteDataSource;

import java.util.ArrayList;
import java.util.List;

class BaseRepository {
    private final Application mApp;
    private final List<RemoteDataSource<?>> mRemoteDataSources;

    BaseRepository(Application app) {
        mApp = app;
        mRemoteDataSources = new ArrayList<>(0);
    }

    final Application getApp() {
        return mApp;
    }

    final void addRemoteDataSource(RemoteDataSource<?> remoteSource) {
        mRemoteDataSources.add(remoteSource);
    }

    public void onUnsubscribe() {
        for (RemoteDataSource<?> remoteDataSource : mRemoteDataSources)
            remoteDataSource.cancel();
    }
}

File: app/src/main/java/com/digis/android/task/implementation/kotlin/data/BaseRepository.kt
package com.digis.android.task.implementation.kotlin.data

import android.app.Application
import com.digis.android.task.implementation.kotlin.data.source.remote.RemoteDataSource

open class BaseRepository(val app: Application) {
    private val remoteDataSources: MutableList<RemoteDataSource<*>> = ArrayList(0)

    protected fun addRemoteDataSource(remoteSource: RemoteDataSource<*>) {
        remoteDataSources.add(remoteSource)
    }

    fun onUnsubscribe() {
        for (remoteDataSource in remoteDataSources)
            remoteDataSource.cancel();
    }
}